DM-38491: Add PydanticRedisStorage and EncryptedPydanticRedisStorage

jonathansick opened the pull request:
These classes are ported from Gafaelfawr, and refactored to allow both unencrypted and encrypted storage. A draft of the docs is at https://safir.lsst.io/v/DM-38491/user-guide/pydantic-redis.html and API docs are at https://safir.lsst.io/v/DM-38491/api/safir.redis.PydanticRedisStorage.html#safir.redis.PydanticRedisStorage

rra reviewed (approved):
This looks great. Thank you so much for doing this! Just a few very minor nits. Looking forward to moving this code out of Gafaelfawr!

jonathansick commented on docs/api.rst:
```diff
@@ -54,6 +54,10 @@ API reference
 .. automodapi:: safir.pydantic
    :include-all-objects:
 
+.. automodapi:: safir.pydanticredis
```
I like this idea. There's not a lot the pydantic-related implemented would add to the safir.redis namespace, so it'd probably work well alongside whatever else we might add in the future.

jonathansick commented:
@rra I went ahead and moved the module to `safir.redis`. I also changed the exception to a SlackException. Does the change and additional tests in https://github.com/lsst-sqre/safir/pull/146/commits/6f879a7c4bf9d224fa92e70135991c04b4cf582a look reasonable to you?

rra commented on src/safir/redis.py:
```diff
@@ -0,0 +1,235 @@
+__all__ = [
+    "PydanticRedisStorage",
+    "EncryptedPydanticRedisStorage",
+    "S",
```
I don't think you want to export this. It's only used by mypy and thus is module local. (Unless this helps the API documentation look nicer, but it looks like you're still excluding it from Sphinx warnings, so presumably Sphinx doesn't yet understand TypeVars.)

jonathansick commented:
I also want to double check that adding `redis` as a default dependency, not an extra, is ok with you. We have extras for things like `sqlalchemy`; I'm not sure whether you'd judge it a burden to install by default or not?

rra commented on src/safir/redis.py:
```diff
@@ -0,0 +1,235 @@
+        try:
+            return self._deserialize(data)
+        except Exception as e:
+            error = f"{type(e).__name__}: {str(e)}"
+            msg = f"Cannot deserialize data for key {full_key}: {error}"
+            raise DeserializeError(msg, key=full_key) from e
```
So here this would look like:
```suggestion
            error = f"{type(e).__name__}: {str(e)}"
            msg = f"Cannot deserialize data for key {full_key}"
            raise DeserializeError(msg, key=full_key, error=error) from e
```

rra commented on src/safir/redis.py:
```diff
@@ -0,0 +1,235 @@
+    possibly decrypted) or deserialized.
+    """
+
+    def __init__(self, msg: str, key: str) -> None:
+        super().__init__(msg)
+        self.key = key
+
+    def to_slack(self) -> SlackMessage:
+        message = super().to_slack()
+        message.fields.append(SlackTextField(heading="Key", text=self.key))
+        return message
```
Because this is Pydantic, which uses newlines and indentation in its exception messages that will be formatted oddly in the main part of the Slack message, I think this would read better in Slack if you did something like this instead:
```suggestion
    def __init__(self, msg: str, key: str, error: str) -> None:
        super().__init__(msg)
        self.key = key
        self.error = error

    def to_slack(self) -> SlackMessage:
        message = super().to_slack()
        message.fields.append(SlackTextField(heading="Key", text=self.key))
        message.blocks.append(SlackCodeBlock(heading="Error", code=self.error))
        return message
```
Then leave the error out of the main exception message. (In the backtrace it will still be captured because the underlying exception will also be logged directly.)

(Maybe this is overkill; if it doesn't sound like a good idea to you, feel free to drop the idea.)

rra reviewed (approved):
This generally looks great, just one formatting tweak that I think will improve the eventual Slack rendering.

rra commented:
Oh, that's a good question on adding Redis as a default dependency. I think I would lean towards making it an option instead. It's not that heavy-weight of a dependency, but I also think most Safir applications won't use Redis (although I dunno, at the rate we're going with Redis deployments, maybe I'm wrong about that!).
